src/charm.py

```diff
@@ -1,29 +1,125 @@
+        # Test the service account we've got for sufficient perms
+        api = kubernetes.client.CoreV1Api(kubernetes.client.ApiClient())
+
+        try:
+            api.list_namespaced_service(namespace=K8sServicePatch.namespace())
+        except kubernetes.client.exceptions.ApiException as e:
+            if e.status == 403:
+                # If we can't read a cluster role, we don't have enough permissions
+                # Admin would need to run `juju trust` on this application to continue
+                # TODO inform caller that `juju trust` is needed
+                return False
+            else:
+                raise e
+        return True
+
+    @staticmethod
+    def _k8s_service(
+        app: str, service_ports: List[Tuple[str, int, int]]
+    ) -> kubernetes.client.V1Service:
+        """Property accessor to return a valid Kubernetes Service representation for Alertmanager
+
+        Args:
+            app: app name
+            service_ports: a list of tuples (name, port, target_port) for every service port.
+
+        Returns:
+            kubernetes.client.V1Service: A Kubernetes Service with correctly annotated metadata and ports
+        """
+
+        ports = [
+            kubernetes.client.V1ServicePort(name=port[0], port=port[1], target_port=port[2])
+            for port in service_ports
+        ]
+
+        ns = K8sServicePatch.namespace()
+        return kubernetes.client.V1Service(
+            api_version="v1",
+            metadata=kubernetes.client.V1ObjectMeta(
+                namespace=ns,
+                name=app,
+                labels={"app.kubernetes.io/name": app},
+            ),
+            spec=kubernetes.client.V1ServiceSpec(
+                ports=ports,
+                selector={"app.kubernetes.io/name": app},
+            ),
+        )
+
+    @staticmethod
+    def patch(app: str, service_ports: List[Tuple[str, int, int]]) -> bool:
+        """Patch the Kubernetes service created by Juju to map the correct port
+
+        Args:
+            app: app name
+            service_ports: a list of tuples (name, port, target_port) for every service port.
+
+        Returns:
+            bool: True indicates successful patch
+        """
+        # First ensure we're authenticated with the Kubernetes API
+        if K8sServicePatch._k8s_auth():
+            ns = K8sServicePatch.namespace()
+            # Set up a Kubernetes client
+            api = kubernetes.client.CoreV1Api(kubernetes.client.ApiClient())
+            # Delete the existing service so we can redefine with correct ports
+            # I don't think you can issue a patch that *replaces* the existing ports,
+            # only append
+            api.delete_namespaced_service(name=app, namespace=ns)
+            # Recreate the service with the correct ports for the application
+            api.create_namespaced_service(
+                namespace=ns, body=K8sServicePatch._k8s_service(app, service_ports)
+            )
```

Comment:
One thing I didn't get around to doing was wrapping this with a try/except and catch some likely exceptions. I can't see a normal situation where this is likely to fail, but nonetheless worth catching so it doesn't just blindly raise. If you see in the auth method, the Kubernetes `APIException` might be the one to catch, and it carries with it some other attributes (like `status` etc.)

Reply:
Both here and in _auth, the immediately obvious exceptions are `ApiTypeError` and `ApiValueError` (neither of which inherit `ApiException`. What could I do under `except` other than `logger.error`?

BTW, all the exception attributes (e.g. `status`, `reason`) are already incorporated into its `__str__`:
```python
    def __str__(self):
        """Custom error messages for exception"""
        error_message = "({0})\n"\
                        "Reason: {1}\n".format(self.status, self.reason)
```